File: 2022/python/day-04.py
def is_contained(startA, endA, startB, endB):
    return int(startB) >= int(startA) and int(endB) <= int(endA) or int(startA) >= int(startB) and int(endA) <= int(endB)

input = open('input-04', 'r').read().strip()
assignements = input.split('\n')
pairs = [assignement.split(',') for assignement in assignements]
ranges = [(range1.split('-'), range2.split('-')) for range1, range2 in pairs]
contained = [is_contained(*range[0], *range[1]) for range in ranges]

print(f'Part1: {sum(contained)}')


def is_overlaping(startA, endA, startB, endB):
    return not (int(endA) < int(startB) or int(endB) < int(startA))

overlaping = [is_overlaping(*range[0], *range[1]) for range in ranges]

print(f'Part2: {sum(overlaping)}')

File: 2022/rust/src/days/day04.rs
use crate::parser::read_unsigned_int;

fn parse_assignement(buffer: &[u8], offset: &mut usize) -> (usize, usize, usize, usize) {
    let start_a = read_unsigned_int(buffer, offset).unwrap();
    *offset += 1;
    let end_a = read_unsigned_int(buffer, offset).unwrap();
    *offset += 1;
    let start_b = read_unsigned_int(buffer, offset).unwrap();
    *offset += 1;
    let end_b = read_unsigned_int(buffer, offset).unwrap();
    *offset += 1;
    return (start_a, end_a, start_b, end_b);
}

pub(crate) fn part1(input: &String) -> String {
    let buffer = input.as_bytes();
    let mut offset = 0;

    let mut contained = 0;
    while offset < buffer.len() {
        let (start_a, end_a, start_b, end_b) = parse_assignement(buffer, &mut offset);
        if start_b >= start_a && end_b <= end_a || start_a >= start_b && end_a <= end_b {
            contained += 1;
        }
    }

    return contained.to_string()
}

pub(crate) fn part2(input: &String) -> String {
    let buffer = input.as_bytes();
    let mut offset = 0;

    let mut overlaping = 0;
    while offset < buffer.len() {
        let (start_a, end_a, start_b, end_b) = parse_assignement(buffer, &mut offset);
        if !(end_a < start_b || end_b < start_a) {
            overlaping += 1;
        }
    }

    return overlaping.to_string()
}